File: SSHSharp/Transport/Session.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using SSHSharp.Verifiers;

namespace SSHSharp.Transport
{
    /// <summary>
    /// The transport layer represents the lowest level of the SSH protocol, and
    /// implements basic message exchanging and protocol initialization. It will
    /// never be instantiated directly (unless you really know what you're about),
    /// but will instead be created for you automatically when you create a new
    /// SSH session via Net::SSH.start.
    /// </summary>
    [Incomplete]
    public class Session
    {
        /// <summary>
        /// The standard port for the SSH protocol.
        /// </summary>
        public const int DefaultPort = 22;

        /// <summary>
        /// The port number to connect to, as given in the options to the constructor.
        /// If no port number was given, this will default to DefaultPort.
        /// </summary>
        public int Port { get; private set; }

        /// <summary>
        /// The host to connect to, as given to the constructor.
        /// </summary>
        public string Host { get; private set; }

        /// <summary>
        /// The underlying socket object being used to communicate with the remote
        /// host.
        /// </summary>
        public NetworkStream Socket { get; private set; }

        /// <summary>
        /// The ServerVersion instance that encapsulates the negotiated protocol
        /// version.
        /// </summary>
        public ServerVersion ServerVersion { get; private set; }

        /// <summary>
        /// The host-key verifier object used to verify host keys, to ensure that
        /// the connection is not being spoofed.
        /// </summary>
        public IHostKeyVerifier HostKeyVerifier { get; private set; }

        /// <summary>
        /// The hash of options that were given to the object at initialization.
        /// </summary>
        public SessionOptions Options { get; set; }

        /// <summary>
        /// The Algorithms instance used to perform key exchanges.
        /// </summary>
        public Algorithms Algorithms { get; private set; }

        /// <summary>
        /// Instantiates a new transport layer abstraction. This will block until
        /// the initial key exchange completes, leaving you with a ready-to-use
        /// transport session.
        /// </summary>
        /// <param name="host"></param>
        public Session(string host) 
            : this(host, new SessionOptions())
        {
        }

        /// <summary>
        /// Instantiates a new transport layer abstraction. This will block until
        /// the initial key exchange completes, leaving you with a ready-to-use
        /// transport session.
        /// </summary>
        /// <param name="host"></param>
        /// <param name="options"></param>
        public Session(string host, SessionOptions options)
        {
            this.Host = host;
            this.Port = options.Port ?? DefaultPort;

            this.Options = options;

            Trace.TraceWarning("establishing connection to {0}:{1}", this.Host, this.Port);
            var factory = options.Proxy ?? ConnectFactory;

            if(options.Timeout != null)
                throw new NotSupportedException("Timeout is not currently supported");

            var socket = Timeout.Wait(options.Timeout ?? 0, () => factory(this.Host, this.Port));

            Trace.TraceInformation("Connection established.");

            this.HostKeyVerifier = SelectHostKeyVerifier(options.Paranoid);

            this.ServerVersion = new ServerVersion(socket);
            
            this.Algorithms = new Algorithms(this, options);

            Wait(() => this.Algorithms.IsInitialized);
        }

        private static PacketStream ConnectFactory(string host, int port)
        {
            var c = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            c.Connect(host, port);
            return new PacketStream(c, true);
        }

        /// <summary>
        /// Waits (blocks) until the given block returns true. If no block is given,
        /// this just waits long enough to see if there are any pending packets. Any
        /// packets read are enqueued (see #push).
        /// </summary>
        /// <param name="func"></param>
        private void Wait(Func<bool> func)
        {
            while (true)
            {
                if (func != null && func())
                    break;

                var message = PollMessage("nonblock", false);
                if (message != null)
                    Push(message);

                if(func == null)
                    break;
            }
        }

        private void Push(object message)
        {
            throw new NotImplementedException();
        }

        private object PollMessage(string nonblock, bool b)
        {
            throw new NotImplementedException();
        }

        public PeerInfo Peer { get; set; }

        private string _hostAsString;
        /// <summary>
        /// Returns the host (and possibly IP address) in a format compatible with
        /// SSH known-host files.
        /// </summary>
        /// <returns></returns>
        public string HostAsString()
        {
            if (_hostAsString != null)
                return _hostAsString;

            var s = this.Host;
            if (this.Port != DefaultPort)
                s = string.Format("[{0}]:{1}", s, this.Port);

            // if socket.peer_ip != host
            if (this.Peer.IPAddress.ToString() != this.Host)
            {
                var s2 = this.Peer.IPAddress.ToString();
                if (this.Port != DefaultPort)
                    s2 = string.Format("[{0}]:{1}", s2, this.Port);

                return s + "," + s2;
            }

            return s;
        }

        /// <summary>
        /// Returns true if the underlying socket has been closed.
        /// </summary>
        public bool IsClosed
        {
            get { return this.Socket.CanRead || this.Socket.CanWrite; }
        }

        /// <summary>
        /// Instantiates a new host-key verification class, based on the value of
        /// the parameter. When true or nil, the default Lenient verifier is
        /// returned. If it is false, the Null verifier is returned, and if it is
        /// :very, the Strict verifier is returned. If the argument happens to
        /// respond to :verify, it is returned directly. Otherwise, an exception
        /// is raised.
        /// </summary>
        /// <param name="paranoid"></param>
        /// <returns></returns>
        private IHostKeyVerifier SelectHostKeyVerifier(object paranoid)
        {
            if (paranoid == null)
                return new Lenient();

            if (paranoid.Equals("very"))
                return new Strict();

            if(paranoid is IHostKeyVerifier)
                return (IHostKeyVerifier) paranoid;

            if(paranoid is bool)
            {
                var b = (bool) paranoid;
                if (b)
                    return new Lenient();
                
                return new Null();
            }

            throw new ArgumentException(string.Format("argument paranoid is not valid: {0}", paranoid), "paranoid");
        }
    }
}

File: net-ssh-2.0.11/lib/net/ssh/transport/session.rb
require 'socket'
require 'timeout'

require 'net/ssh/loggable'
require 'net/ssh/transport/algorithms'
require 'net/ssh/transport/constants'
require 'net/ssh/transport/packet_stream'

module Net; module SSH; module Transport

  
  class Session
    include Constants, Loggable

   
    

   

   

    # Returns the host (and possibly IP address) in a format compatible with
    # SSH known-host files.
    def host_as_string
      @host_as_string ||= begin
        string = "#{host}"
        string = "[#{string}]:#{port}" if port != DEFAULT_PORT
        if socket.peer_ip != host
          string2 = socket.peer_ip
          string2 = "[#{string2}]:#{port}" if port != DEFAULT_PORT
          string << "," << string2
        end
        string
      end
    end

    # Returns true if the underlying socket has been closed.
    def closed?
      socket.closed?
    end

    # Cleans up (see PacketStream#cleanup) and closes the underlying socket.
    def close
      socket.cleanup
      socket.close
    end

    # Performs a "hard" shutdown of the connection. In general, this should
    # never be done, but it might be necessary (in a rescue clause, for instance,
    # when the connection needs to close but you don't know the status of the
    # underlying protocol's state).
    def shutdown!
      error { "forcing connection closed" }
      socket.close
    end

    # Returns a new service_request packet for the given service name, ready
    # for sending to the server.
    def service_request(service)
      Net::SSH::Buffer.from(:byte, SERVICE_REQUEST, :string, service)
    end

    # Requests a rekey operation, and blocks until the operation completes.
    # If a rekey is already pending, this returns immediately, having no
    # effect.
    def rekey!
      if !algorithms.pending?
        algorithms.rekey!
        wait { algorithms.initialized? }
      end
    end

    # Returns immediately if a rekey is already in process. Otherwise, if a
    # rekey is needed (as indicated by the socket, see PacketStream#if_needs_rekey?)
    # one is performed, causing this method to block until it completes.
    def rekey_as_needed
      return if algorithms.pending?
      socket.if_needs_rekey? { rekey! }
    end

    # Returns a hash of information about the peer (remote) side of the socket,
    # including :ip, :port, :host, and :canonized (see #host_as_string).
    def peer
      @peer ||= { :ip => socket.peer_ip, :port => @port.to_i, :host => @host, :canonized => host_as_string }
    end

    # Blocks until a new packet is available to be read, and returns that
    # packet. See #poll_message.
    def next_message
      poll_message(:block)
    end

    # Tries to read the next packet from the socket. If mode is :nonblock (the
    # default), this will not block and will return nil if there are no packets
    # waiting to be read. Otherwise, this will block until a packet is
    # available. Note that some packet types (DISCONNECT, IGNORE, UNIMPLEMENTED,
    # DEBUG, and KEXINIT) are handled silently by this method, and will never
    # be returned.
    #
    # If a key-exchange is in process and a disallowed packet type is
    # received, it will be enqueued and otherwise ignored. When a key-exchange
    # is not in process, and consume_queue is true, packets will be first
    # read from the queue before the socket is queried.
    def poll_message(mode=:nonblock, consume_queue=true)
      loop do
        if consume_queue && @queue.any? && algorithms.allow?(@queue.first)
          return @queue.shift
        end

        packet = socket.next_packet(mode)
        return nil if packet.nil?

        case packet.type
        when DISCONNECT
          raise Net::SSH::Disconnect, "disconnected: #{packet[:description]} (#{packet[:reason_code]})"

        when IGNORE
          debug { "IGNORE packet recieved: #{packet[:data].inspect}" }

        when UNIMPLEMENTED
          lwarn { "UNIMPLEMENTED: #{packet[:number]}" }

        when DEBUG
          send(packet[:always_display] ? :fatal : :debug) { packet[:message] }

        when KEXINIT
          algorithms.accept_kexinit(packet)

        else
          return packet if algorithms.allow?(packet)
          push(packet)
        end
      end
    end



    # Adds the given packet to the packet queue. If the queue is non-empty,
    # #poll_message will return packets from the queue in the order they
    # were received.
    def push(packet)
      @queue.push(packet)
    end

    # Sends the given message via the packet stream, blocking until the
    # entire message has been sent.
    def send_message(message)
      socket.send_packet(message)
    end

    # Enqueues the given message, such that it will be sent at the earliest
    # opportunity. This does not block, but returns immediately.
    def enqueue_message(message)
      socket.enqueue_packet(message)
    end

    # Configure's the packet stream's client state with the given set of
    # options. This is typically used to define the cipher, compression, and
    # hmac algorithms to use when sending packets to the server.
    def configure_client(options={})
      socket.client.set(options)
    end

    # Configure's the packet stream's server state with the given set of
    # options. This is typically used to define the cipher, compression, and
    # hmac algorithms to use when reading packets from the server.
    def configure_server(options={})
      socket.server.set(options)
    end

    # Sets a new hint for the packet stream, which the packet stream may use
    # to change its behavior. (See PacketStream#hints).
    def hint(which, value=true)
      socket.hints[which] = value
    end

    public

      # this method is primarily for use in tests
      attr_reader :queue #:nodoc:


  end
end; end; end
